### PR TITLE
add bf16 support for bucketize

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61015 add BFloat16 support for BatchNorm2d on CPU
* #58297 add BFloat16 support for UpSample on CPU
* #56904 optimize transpose copy for float32 and bfloat16 on CPU
* #56903 add BFloat16 support for MaxPool2d on CPU
* #56902 add BFloat16 support for AdaptiveAvgPool2d on CPU
* #56372 add BFloat16 support for bernoulli and Dropout on CPU
* **#55588 add bf16 support for bucketize**
* #55221 optimize BFloat16 elemwise operators CPU: sigmoid, sigmoid_backward, tanh_backward, addcmul, addcdiv

Differential Revision: [D28836796](https://our.internmc.facebook.com/intern/diff/D28836796)

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836796).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836796).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836796).

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
Looks good, please include benchmarks comparing float and bfloat16 implementations

**Comment by mingfeima** [comment]
Updates: add performance compare for float32 and bfloat16.

#### avx512 CPU single core:
```bash
bucketize, dtype = torch.float32 ; size:  128 x 1024 ; time = 2.109 ms
bucketize, dtype = torch.bfloat16 ; size:  128 x 1024 ; time = 2.227 ms
```

#### avx2 CPU single core:
```bash
bucketize, dtype = torch.float32 ; size:  128 x 1024 ; time = 2.728 ms
bucketize, dtype = torch.bfloat16 ; size:  128 x 1024 ; time = 2.786 ms
```

This PR has no change on float32 performance.

**Comment by JamesDeAntonis** [comment]
I'm involved in [this huggingface pr](https://github.com/huggingface/transformers/issues/13170) to add bf16 support to their trainer.  I'm mostly done, but my benchmarking has shown that it's actually slower than float32.  By comparison, fp16 is much faster than float32.  My profiling investigation shows that this is because of the backward pass.  Do you think this pr addresses this performance issue?

As a follow-up, @mingfeima's benchmarking looks like bf16 does not do better than float32?  In theory it should be much faster right?

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28836796).

**Comment by mingfeima** [comment]
@JamesDeAntonis, `BFloat16` is 'always' faster than `float32` is not a correct recognition.

bfloat16 is more like a storage type rather a data type, similar to int8 case. From the performance wise:
* For operators where we can have native instruction support from hardware (intel Xeon CPU, sapphire rapids) with [AMX](https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/intrinsics/intrinsics-for-intel-advanced-matrix-extensions-intel-amx-instructions.html) or AVX-512_BF16, bfloat16 is going to be significantly faster than float32. `convolution` and `gemm` is of this case, and the interface is via oneDNN (intel's acceleration library).
* For the rest of operators, bfloat16 needs to be converted to float32 for following computations, since bfloat16 is actually stored as int16, the compiler would not generate correct result if you directly multiple two int16. And the result needs to be converted from float32 back to bfloat16.

The data type conversion is time consuming but we can vectorize it. And bfloat16 has 1/2 memory footprint v.s. float32 (less memory access). SO whether it is going to be faster or slower is a balance between data type conversion and the memory access consumption, which can be:
* Memory bandwidth bound operators -  faster. e.g. element-wise operations.
* Compute bound operators - slower. e.g. Softmax. (softmax has exp which needs taylor expansion).
* Scalar only operators - slower. This is the worst case as we can't vectorize the data type conversion. Bucketize is this case.

From the big picture, bfloat16 is surely profitable, for instance you can speedup the Block of `conv-bn-relu` using bfloat16 and leave the rest (which might be slower) with float32, with the help of `torch.autocast`. So we need to make sure the *major* blocks has performance benefit with bfloat16.

We need to enrich the bfloat16 coverage not because bfloat16 is 100% faster than float32, but to maintain the dtype propagate in pytorch. e.g. `a - b - c`, if we run `a` in bfloat16, and `b` or `c` don't have bfloat16 support, it is going to be runtime error.
